### PR TITLE
Fixes #19440: Directive  tree API

### PR DESCRIPTION
**Pull request by VinceMacBuche**
https://issues.rudder.io/issues/19440

### CODE REVIEW

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/EndpointsDefinition.scala [hunk 1]
```diff
@@ -171,6 +171,10 @@ object DirectiveApi extends ApiModuleProvider[DirectiveApi] {
     val description = "List all directives"
     val (action, path)  = GET / "directives"
   }
+  final case object DirectiveTree extends DirectiveApi with ZeroParam with StartsAtVersion14 with SortIndex { val z = implicitly[Line].value
+    val description = "Update given directive information"
```
Comment (fanf) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/EndpointsDefinition.scala:175]: the description is not correct

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/DirectiveApi.scala [hunk 2]
```diff
@@ -529,14 +536,26 @@ class DirectiveApiService14 (
...
   , workflowLevelService : WorkflowLevelService
-  , restExtractor        : RestExtractorService
   , editorService        : DirectiveEditorService
   , restDataSerializer   : RestDataSerializer
   , techniqueRepository  : TechniqueRepository
   ) {
 
   def serialize = restDataSerializer.serializeDirective _
 
+  def directiveTree() : IOResult[JRDirectiveTreeCategory] = {
```
Comment (fanf) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/DirectiveApi.scala:546]: Can you add a comment that system techniques are excluded ?

**Review by fanf — COMMENTED** [review]
Can you add a test in `webapp/sources/rudder/rudder-rest/src/test/resources/api/api_directives.yml` ?

**Comment by VinceMacBuche** [comment]
PR updated with a new commit

**Comment by VinceMacBuche** [comment]
PR updated with a new commit

**Review by fanf — COMMENTED** [review]
Just missing a test in `webapp/sources/rudder/rudder-rest/src/test/resources/api/api_directives.yml` and can be merged

**Comment by VinceMacBuche** [comment]
PR updated with a new commit

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/DirectiveApi.scala [hunk 3]
```diff
@@ -551,8 +551,8 @@ class DirectiveApiService14 (
   def directiveTree(includeSystem : Boolean) : IOResult[JRDirectiveTreeCategory] = {
     def filterSystem(cat : FullActiveTechniqueCategory) : FullActiveTechniqueCategory = {
       cat.copy(
-          subCategories    = cat.subCategories.filter(c => includeSystem || c.isSystem).map(filterSystem)
-        , activeTechniques = cat.activeTechniques.filterNot(c => includeSystem || c.isSystem).map(t => t.copy(directives = t.directives.filterNot(_.isSystem)))
+          subCategories    = cat.subCategories.filter(c => includeSystem || !c.isSystem).map(filterSystem)
```
Comment (fanf) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/DirectiveApi.scala:554]: It should be sorted too, to be consistent between run and help the user who is calling it to see where things are

Comment (fanf) [reply webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/DirectiveApi.scala:554]: sorted by category and technique human readable name

**Review by fanf — COMMENTED** [review]
It's ok with me. Happy the test helped found a bug. 
You can self merge once you are satisfied with the result.

**Comment by VinceMacBuche** [comment]
PR updated with a new commit

**Pull request merged**